### DISCUSSION TITLE
Unprocessable entity error with optional str enum set

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#8663] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional, Set

from fastapi import FastAPI, Query
from enum import Enum


class Country(str, Enum):
    eu = "eu"
    us = "us"
    cn = "cn"
    ru = "ru"


app = FastAPI()


@app.get("/")
def get_something(country: Optional[Set[Country]] = Query(None)):
    return {"country": country}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="127.0.0.1", port=9000)
```
![image](https://user-images.githubusercontent.com/79832454/138901660-de85cfb3-9491-43e2-80c4-fcda2cab3134.png)

### Description

Having a parameter that is `None` with `Optional[Set[Country]] = Query(None)` creates an incorrect curl request and therefore a 422 HTTP error. This might be a SwaggerUI problem, because removing the parameter in the curl command solves the problem.

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

3.8.10

**Comment by AlexandreBiguet** [comment, marked as answer]
Hi :)

The response body associated with the request is

```json
{
  "detail": [
    {
      "loc": [
        "query",
        "country",
        0
      ],
      "msg": "value is not a valid enumeration member; permitted: 'eu', 'us', 'cn', 'ru'",
      "type": "type_error.enum",
      "ctx": {
        "enum_values": [
          "eu",
          "us",
          "cn",
          "ru"
        ]
      }
    }
  ]
}
```

The error comes from the fact that there is no `""` value for the Enum, but the curl request is not incorrect. 

For example, if you add another endpoint to your app

```python

from typing import Optional, Set, Any

from fastapi import FastAPI, Query
from enum import Enum


class Country(str, Enum):
    eu = "eu"
    us = "us"
    cn = "cn"
    ru = "ru"


app = FastAPI()


@app.get("/")
def get_something(country: Optional[Set[Country]] = Query(None)):
    return {"country": country}


@app.get("/other")
def get_something_else(data: Optional[Set[Any]] = Query(None)):
    return {"data": data}
```

you would be able to run the following curl request:

```bash
➜  ~ curl -X 'GET' \
  'http://127.0.0.1:8000/other?data=' \    
  -H 'accept: application/json'
{"data":[""]}
```
You might want to define the Enum like this

```python
class Country(str, Enum):
    eu = "eu"
    us = "us"
    cn = "cn"
    ru = "ru"
    default = ""
```

which would allow 
```bash
➜  ~ curl -X 'GET' \
  'http://127.0.0.1:8000/?country=' \
  -H 'accept: application/json'
{"country":[""]}
```

---


If you want the enum to return a default value when the given value does not exist, it is possible to override the `_missing_` method:

```python
class Country(str, Enum):
    eu = "eu"
    us = "us"
    cn = "cn"
    ru = "ru"

    @classmethod
    def _missing_(cls, value):
        return Country.eu

```
In that case, the curl request would return:
```bash
➜  ~ curl -X 'GET' \
  'http://127.0.0.1:8000/?country=' \    
  -H 'accept: application/json'
{"country":["eu"]}
```

There is a drawback of doing this though is that your user will be able to 

```bash
➜  ~ curl -X 'GET' \
  'http://127.0.0.1:8000/?country=you%20can%20write%20what%20you%20want%20here\"' \
  -H 'accept: application/json'
{"country":["eu"]}%  
```
which is probably not something that you want. 

Hope this helps, 

Cheers :)
